### PR TITLE
[Shamir port] Implement greeting procedure

### LINKED ISSUE #8841: [Shamir port] Implement greeting procedure
Part of #6090

### PR DESCRIPTION
**Pull request by vxgmichel**
Close #8841

TODO:
- [x] more tests
- [x] proper error handling
- [x] greeter detects when the device is not recoverable due to too many recipients revoked
- [x] claimer detects when the device is not recoverable due to too many recipients being revoked

### CODE REVIEW

File: libparsec/crates/client/tests/unit/invite/shamir.rs [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+use libparsec_types::prelude::*;
+
+use crate::{
+    claimer_retrieve_info, ClientConfig, MountpointMountStrategy, ProxyConfig,
+    ShamirRecoveryClaimMaybeRecoverDeviceCtx, UserOrDeviceClaimInitialCtx,
+    WorkspaceStorageCacheSize,
+};
+
+#[parsec_test(testbed = "shamir", with_server)]
+async fn shamir(tmp_path: TmpPath, env: &TestbedEnv) {
```
Comment (FirelightFlagboy) [inline libparsec/crates/client/tests/unit/invite/shamir.rs:17]: This function is long, I would split it using sub-function

File: cli/src/commands/invite/claim.rs [hunk 2]
```diff
@@ -64,6 +64,9 @@ pub async fn main(args: Args) -> anyhow::Result<()> {
             let ctx = step4_device(ctx).await?;
             save_device(ctx, save_mode).await
         }
+        UserOrDeviceClaimInitialCtx::ShamirRecovery(_) => {
+            panic!("Shamir recovery invitation is not supported")
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply cli/src/commands/invite/claim.rs:74]: Yes, it's part of the todo list

File: libparsec/crates/client/src/invite/claimer.rs [hunk 3]
```diff
@@ -263,20 +270,229 @@ pub async fn claimer_retrieve_info(
...
+
+        let mut recovery_device =
+            LocalDevice::decrypt_and_load(&ciphered_data, &self.secret.data_key)
+                .map_err(|e| ClaimInProgressError::Internal(anyhow::Error::msg(e)))?;
+
+        // When using the tested, the recovery device organization address is set to a placeholder address.
+        // This is because the organization address is not known yet when the testbed is initialized.
+        // In this case, we replace the placeholder address with the actual organization address.
+        if cfg!(test)
+            && recovery_device
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply libparsec/crates/client/src/invite/claimer.rs:447]: This suggestion would complain about `recovery_device` being `mut`

File: libparsec/crates/client/src/certif/shamir_recovery_list.rs [hunk 4]
```diff
@@ -331,3 +331,40 @@ pub async fn list_shamir_recoveries_for_others(
...
         .await?
 }
+
+pub async fn get_shamir_recovery_share_data(
+    ops: &CertificateOps,
+    user_id: UserID,
+) -> Result<ShamirRecoveryShareData, CertifListShamirRecoveryError> {
+    ops.store
+        .for_read(|store| async move {
+            // TODO: check that the shamir is actually recoverable
```
Comment (AureliaDolo) [inline libparsec/crates/client/src/certif/shamir_recovery_list.rs:341]: Meaning that enough recipient are still here / not revoked ?

File: libparsec/crates/protocol/schema/invited_cmds/invite_info.json5
Comment (AureliaDolo) [inline libparsec/crates/protocol/schema/invited_cmds/invite_info.json5:1]: Update rfc accordingly

File: libparsec/crates/client/src/certif/shamir_recovery_list.rs (hunk 4 above)
Comment (vxgmichel) [reply libparsec/crates/client/src/certif/shamir_recovery_list.rs]: Exactly

File: libparsec/crates/client/tests/unit/invite/shamir.rs (hunk 1 above)
Comment (vxgmichel) [reply libparsec/crates/client/tests/unit/invite/shamir.rs:17]: I'm not convinced this simplifies the reading :)

File: libparsec/crates/client/src/invite/claimer.rs [hunk 5]
```diff
@@ -207,9 +218,10 @@ async fn run_claimer_step_until_ready(
 }
 
 #[derive(Debug)]
-pub enum UserOrDeviceClaimInitialCtx {
+pub enum AnyClaimRetrievedInfoCtx {
     User(UserClaimInitialCtx),
     Device(DeviceClaimInitialCtx),
+    ShamirRecovery(ShamirRecoveryClaimPickRecipientCtx),
```
Comment (touilleMan) [inline libparsec/crates/client/src/invite/claimer.rs:224]: So in the end no `ShamirRecoveryNonRecoverable(ShamirRecoveryNonRecoverableCtx)` variant, we stick with the `is_recoverable` method instead ?

**Review by touilleMan — CHANGES_REQUESTED** [review]
tests for `Client::start_shamir_recovery_invitation_greet` are missing (but that's basically just renaming `libparsec/crates/client/tests/unit/client/shamir_recovery_get_share_data.rs`)

File: server/parsec/components/memory/invite.py [hunk 6]
```diff
@@ -566,6 +568,40 @@ async def info_as_invited(
...
+        if invitation.claimer_user_id is None:
+            return InviteShamirRecoveryRevealBadOutcome.DATA_NOT_FOUND
+
+        shamir_recoveries = org.shamir_recoveries.get(invitation.claimer_user_id, [])
+        if not shamir_recoveries:
+            return InviteShamirRecoveryRevealBadOutcome.DATA_NOT_FOUND
+
+        *_, shamir_recovery = shamir_recoveries
+        if shamir_recovery.reveal_token != reveal_token:
+            return InviteShamirRecoveryRevealBadOutcome.DATA_NOT_FOUND
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply server/parsec/components/memory/invite.py]: 56e035bb5077545fccdf300d714dcee7c03d26f3

File: libparsec/crates/protocol/schema/invited_cmds/invite_shamir_recovery_reveal.json5 [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+                "status": "ok",
+                "fields": [
+                    {
+                        "name": "ciphered_data",
+                        "type": "Bytes"
+                    }
+                ]
+            },
+            {
+                "status": "not_found"
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply libparsec/crates/protocol/schema/invited_cmds/invite_shamir_recovery_reveal.json5]: 56e035bb5077545fccdf300d714dcee7c03d26f3

File: misc/gen_protocol_typings.py [hunk 8]
```diff
@@ -98,6 +98,8 @@ def cook_field_type(
         ("Size", "int"),
         ("Index", "int"),
         ("NonZeroInteger", "int"),
+        ("NonZeroU8", "int"),
+        ("IntegerBetween1And100", "int"),
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply misc/gen_protocol_typings.py]: 56e035bb5077545fccdf300d714dcee7c03d26f3

File: server/parsec/components/memory/invite.py [hunk 9]
```diff
@@ -566,6 +568,40 @@ async def info_as_invited(
...
+        except KeyError:
+            return InviteShamirRecoveryRevealBadOutcome.INVITATION_NOT_FOUND
+        if invitation.is_deleted:
+            return InviteShamirRecoveryRevealBadOutcome.INVITATION_DELETED
+
+        if invitation.claimer_user_id is None:
+            return InviteShamirRecoveryRevealBadOutcome.DATA_NOT_FOUND
+
+        shamir_recoveries = org.shamir_recoveries.get(invitation.claimer_user_id, [])
+        if not shamir_recoveries:
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply server/parsec/components/memory/invite.py:601]: 56e035bb5077545fccdf300d714dcee7c03d26f3

File: libparsec/crates/client/tests/unit/invite/shamir.rs [hunk 10]
```diff
@@ -0,0 +1,335 @@
...
+        libparsec_types::InvitationType::User,
+        alice_token,
+    );
+
+    // Bob delete the invitation
+    let bob_client = client_factory(&env.discriminant_dir, bob.clone()).await;
+    bob_client.cancel_invitation(alice_token).await.unwrap();
+
+    let error = claimer_retrieve_info(config, addr, None).await.unwrap_err();
+    p_assert_matches!(&error, ClaimerRetrieveInfoError::AlreadyUsed);
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply libparsec/crates/client/tests/unit/invite/shamir.rs:334]: This error variant in `ClaimInProgressError` was already called `AlreadyUsed`, which is exposed in the bindings.

Should I rename them both anyway?

File: libparsec/crates/client/tests/unit/invite/shamir.rs [hunk 11]
```diff
@@ -0,0 +1,335 @@
+// Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 2016-present Scille SAS
+
+use std::{collections::HashMap, sync::Arc};
+
+use super::super::super::client::tests::utils::client_factory;
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply libparsec/crates/client/tests/unit/invite/shamir.rs]: af79f921b7d831fd34b0151191d50a97996a280f

File: libparsec/crates/client/tests/unit/invite/shamir.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (touilleMan) [reply libparsec/crates/client/tests/unit/invite/shamir.rs:334]: yes, but you can do that in another PR if you prefer ^^

File: libparsec/crates/client/src/client/shamir_recovery_list.rs [hunk 12]
```diff
@@ -1,23 +1,36 @@
...
     client: &Client,
-) -> Result<Vec<OtherShamirRecoveryInfo>, ClientListShamirRecoveryError> {
+) -> Result<Vec<OtherShamirRecoveryInfo>, ClientListShamirRecoveriesForOthersError> {
     client
         .certificates_ops
         .list_shamir_recoveries_for_others()
         .await
 }
+
+pub async fn get_shamir_recovery_share_data(
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply libparsec/crates/client/src/client/shamir_recovery_list.rs]: 9ca2567947a0bf90e37798a51e7027e4e2d6ed92

File: libparsec/crates/client/src/invite/claimer.rs [hunk 13]
```diff
@@ -263,20 +275,304 @@ pub async fn claimer_retrieve_info(
...
+        share_ctx: ShamirRecoveryClaimShare,
+    ) -> Result<ShamirRecoveryClaimMaybeRecoverDeviceCtx, ShamirRecoveryClaimAddShareError> {
+        let mut shares = self.shares;
+        // Note that we do not check if the share is already present
+        // This is to avoid handling an extra error that has no real value,
+        // since adding shares can be seen as an idempotent operation.
+        shares.insert(share_ctx.recipient, share_ctx.weighted_share);
+        if shares
+            .values()
+            .map(|shares| shares.len() as u64)
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply libparsec/crates/client/src/invite/claimer.rs]: 054442761078109813c034b1fd6d93c0c12dbc22

File: libparsec/crates/client/src/invite/claimer.rs [hunk 14]
```diff
@@ -1046,3 +1503,68 @@ impl DeviceClaimFinalizeCtx {
...
+
+            let cmds = match AuthenticatedCmds::new(
+                &self.config.config_dir,
+                self.new_local_device.clone(),
+                self.config.proxy.clone(),
+            ) {
+                Ok(cmds) => cmds,
+                Err(e) => {
+                    log::error!("Error while creating authenticated commands: {:?}", e);
+                    return Ok(available_device);
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply libparsec/crates/client/src/invite/claimer.rs]: 054442761078109813c034b1fd6d93c0c12dbc22

File: libparsec/crates/client/src/certif/shamir_recovery_list.rs [hunk 15]
```diff
@@ -331,3 +348,115 @@ pub async fn list_shamir_recoveries_for_others(
...
+                .get_last_shamir_recovery_share_certificate_for_recipient(
+                    UpTo::Current,
+                    user_id,
+                    ops.device.user_id,
+                )
+                .await?;
+            let share_certificate = match share_certificate {
+                Some(share_certificate) => Ok(share_certificate),
+                None => {
+                    Err(CertifGetShamirRecoveryShareDataError::ShamirRecoveryNotFound { user_id })
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply libparsec/crates/client/src/certif/shamir_recovery_list.rs]: `ShamirRecoveryShareNotFound` but yes:
4675850abb2c488e333026137233cb05774cacd1

File: libparsec/crates/client/src/certif/shamir_recovery_list.rs [hunk 16]
```diff
@@ -331,3 +348,115 @@ pub async fn list_shamir_recoveries_for_others(
...
+            let author_verify_key = store
+                .get_device_verify_key(UpTo::Current, share_certificate.author)
+                .await
+                .map_err(|e| match e {
+                    GetCertificateError::NonExisting => {
+                        CertifGetShamirRecoveryShareDataError::VerifyKeyNotFound {
+                            device_id: share_certificate.author,
+                        }
+                    }
+                    GetCertificateError::ExistButTooRecent {
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply libparsec/crates/client/src/certif/shamir_recovery_list.rs]: 4675850abb2c488e333026137233cb05774cacd1

File: libparsec/crates/client/tests/unit/invite/shamir.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (vxgmichel) [reply libparsec/crates/client/tests/unit/invite/shamir.rs]: eec7469da34970848861a591d5418c87b9f2968e

File: libparsec/crates/client/src/invite/claimer.rs (hunk 5 above)
Comment (vxgmichel) [reply libparsec/crates/client/src/invite/claimer.rs:224]: Yes, easier this way

**Review by FirelightFlagboy — DISMISSED** [review]

File: cli/src/commands/invite/greet.rs
Comment (FirelightFlagboy) [inline cli/src/commands/invite/greet.rs:1]: Could be revert now

File: libparsec/crates/client/src/client/mod.rs [hunk 17]
```diff
@@ -591,4 +613,4 @@ impl Client {
 #[cfg(test)]
 #[path = "../../tests/unit/client/mod.rs"]
 #[allow(clippy::unwrap_used)]
-mod tests;
+pub mod tests;
```
Comment (touilleMan) [inline libparsec/crates/client/src/client/mod.rs:616]: pub is no longer needed

File: libparsec/crates/client/src/certif/shamir_recovery_list.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (vxgmichel) [reply libparsec/crates/client/src/certif/shamir_recovery_list.rs]: f88786d51

File: libparsec/crates/protocol/tests/invited_cmds/v4/invite_info.rs [hunk 18]
```diff
@@ -82,23 +82,24 @@ pub fn rep_ok() {
             }),
         ),
         (
-            // Generated from Parsec 3.1.1-a.0+dev
+            // Generated from Parsec 3.2.1-a.0+dev
```
[the remark this answers is not in the document]
Comment (vxgmichel) [reply libparsec/crates/protocol/tests/invited_cmds/v4/invite_info.rs:85]: Good point in general, although in that case this test case is specifically for a shamir recovery invitation, which is a new feature. Hence, there is no compatibility to maintain.

**Pull request merged**